File: lib/lim/src/logical/Package.cpp
/*
 *  This file is part of OpenStaticAnalyzer.
 *
 *  Copyright (c) 2004-2018 Department of Software Engineering - University of Szeged
 *
 *  Licensed under Version 1.2 of the EUPL (the "Licence");
 *
 *  You may not use this work except in compliance with the Licence.
 *
 *  You may obtain a copy of the Licence in the LICENSE file or at:
 *
 *  https://joinup.ec.europa.eu/software/page/eupl
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the Licence is distributed on an "AS IS" basis,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the Licence for the specific language governing permissions and
 *  limitations under the Licence.
 */

#include "lim/inc/lim.h"
#include "lim/inc/Common.h"
#include "common/inc/WriteMessage.h"

#include "lim/inc/messages.h"
#include <algorithm>
#include <string.h>
#include "common/inc/math/common.h"


namespace columbus { namespace lim { namespace asg {

typedef boost::crc_32_type  Crc_type;

namespace logical { 
  Package::Package(NodeId _id, Factory *_factory) :
    Scope(_id, _factory),
    m_packageKind(pkPackage)
  {
  }

  Package::~Package() {
  }

  void Package::clone(const Package& other, bool tryOnVirtualParent) {
    logical::Scope::clone(other, false);

    m_packageKind = other.m_packageKind;
  }

  void Package::prepareDelete(bool tryOnVirtualParent){
    logical::Scope::prepareDelete(false);
  }

  NodeKind Package::getNodeKind() const {
    return ndkPackage;
  }

  PackageKind Package::getPackageKind() const {
    return m_packageKind;
  }

  void Package::setPackageKind(PackageKind _packageKind) {
    m_packageKind = _packageKind;
  }

  bool Package::setEdge(EdgeKind edgeKind, NodeId edgeEnd, void* acValue, bool tryOnVirtualParent) {
    if (logical::Scope::setEdge(edgeKind, edgeEnd, acValue, false)) {
      return true;
    }
    return false;
  }

  bool Package::removeEdge(EdgeKind edgeKind, NodeId edgeEnd, bool tryOnVirtualParent) {
    if (logical::Scope::removeEdge(edgeKind, edgeEnd, false)) {
      return true;
    }
    return false;
  }

  void Package::accept(Visitor &visitor) const {
    visitor.visit(*this);
  }

  void Package::acceptEnd(Visitor &visitor) const {
    visitor.visitEnd(*this);
  }

  double Package::getSimilarity(const base::Base& base){
    if(base.getNodeKind() == getNodeKind()) {
      const Package& node = dynamic_cast<const Package&>(base);
      double matchAttrs = 0;
      std::string str1, str2;
      size_t strMax;
      double strSim;
      str1 = getName();
      str2 = node.getName();
      strMax = std::max(str1.size(), str2.size());
      strSim = 1 - ((double)common::math::editDistance(str1, str2) / (strMax > 0 ? strMax : 1));
      if (strSim < Common::SimilarityMinForStrings)
        return 0.0;
      matchAttrs += strSim;
      if(node.getAccessibility() == getAccessibility()) ++matchAttrs;
      if(node.getCommentLines() == getCommentLines()) ++matchAttrs;
      str1 = getDemangledName();
      str2 = node.getDemangledName();
      strMax = std::max(str1.size(), str2.size());
      strSim = 1 - ((double)common::math::editDistance(str1, str2) / (strMax > 0 ? strMax : 1));
      if (strSim < Common::SimilarityMinForStrings)
        return 0.0;
      matchAttrs += strSim;
      if(node.getIsCompilerGenerated() == getIsCompilerGenerated()) ++matchAttrs;
      if(node.getIsStatic() == getIsStatic()) ++matchAttrs;
      if(node.getLanguage() == getLanguage()) ++matchAttrs;
      str1 = getMangledName();
      str2 = node.getMangledName();
      strMax = std::max(str1.size(), str2.size());
      strSim = 1 - ((double)common::math::editDistance(str1, str2) / (strMax > 0 ? strMax : 1));
      if (strSim < Common::SimilarityMinForStrings)
        return 0.0;
      matchAttrs += strSim;
      if(node.getLLOC() == getLLOC()) ++matchAttrs;
      if(node.getLOC() == getLOC()) ++matchAttrs;
      if(node.getTLLOC() == getTLLOC()) ++matchAttrs;
      if(node.getTLOC() == getTLOC()) ++matchAttrs;
      if(node.getIsAnonymous() == getIsAnonymous()) ++matchAttrs;
      if(node.getPackageKind() == getPackageKind()) ++matchAttrs;
      return matchAttrs / (14 / (1 - Common::SimilarityMinimum)) + Common::SimilarityMinimum;
    } else {
      return 0.0;
    }
  }

  void Package::swapStringTable(RefDistributorStrTable& newStrTable, std::map<Key,Key>& oldAndNewStrKeyMap ){
    std::map<Key,Key>::iterator foundKeyId;

    foundKeyId = oldAndNewStrKeyMap.find(m_demangledName);
    if (foundKeyId != oldAndNewStrKeyMap.end()) {
      m_demangledName = foundKeyId->second;
    } else {
      Key oldkey = m_demangledName;
      m_demangledName = newStrTable.set(factory->getStringTable().get(m_demangledName));
      oldAndNewStrKeyMap.insert(std::pair<Key,Key>(oldkey,m_demangledName));    }

    foundKeyId = oldAndNewStrKeyMap.find(m_mangledName);
    if (foundKeyId != oldAndNewStrKeyMap.end()) {
      m_mangledName = foundKeyId->second;
    } else {
      Key oldkey = m_mangledName;
      m_mangledName = newStrTable.set(factory->getStringTable().get(m_mangledName));
      oldAndNewStrKeyMap.insert(std::pair<Key,Key>(oldkey,m_mangledName));    }

    foundKeyId = oldAndNewStrKeyMap.find(m_name);
    if (foundKeyId != oldAndNewStrKeyMap.end()) {
      m_name = foundKeyId->second;
    } else {
      Key oldkey = m_name;
      m_name = newStrTable.set(factory->getStringTable().get(m_name));
      oldAndNewStrKeyMap.insert(std::pair<Key,Key>(oldkey,m_name));    }

  }

  NodeHashType Package::getHash(std::set<NodeId>& travNodes) const {
    if (hashOk) return nodeHashCache;
    common::WriteMsg::write(CMSG_GET_THE_NODE_HASH_OF_NODE_BEGIN,this->getId()); 
    if (travNodes.count(getId())>0) {
      common::WriteMsg::write(CMSG_GET_THE_NODE_HASH_OF_NODE_SKIP);
      return 0;
    }
    travNodes.insert(getId());
    Crc_type resultHash;
    resultHash.process_bytes( "logical::Package", strlen("logical::Package"));
    common::WriteMsg::write(CMSG_GET_THE_NODE_HASH_OF_NODE_END,resultHash.checksum()); 
    nodeHashCache = resultHash.checksum();
    hashOk = true;
    return nodeHashCache;
  }

  void Package::save(io::BinaryIO &binIo,bool withVirtualBase  /*= true*/) const {
    Scope::save(binIo,false);

    binIo.writeUByte1(m_packageKind);

  }

  void Package::load(io::BinaryIO &binIo, bool withVirtualBase /*= true*/) {
    Scope::load(binIo,false);

    m_packageKind = (PackageKind)binIo.readUByte1();

  }


}


}}}

File: java/lib/java/src/main/java/columbus/java/asg/struc/Package.java
/*
 *  This file is part of OpenStaticAnalyzer.
 *
 *  Copyright (c) 2004-2018 Department of Software Engineering - University of Szeged
 *
 *  Licensed under Version 1.2 of the EUPL (the "Licence");
 *
 *  You may not use this work except in compliance with the Licence.
 *
 *  You may obtain a copy of the Licence in the LICENSE file or at:
 *
 *  https://joinup.ec.europa.eu/software/page/eupl
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the Licence is distributed on an "AS IS" basis,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the Licence for the specific language governing permissions and
 *  limitations under the Licence.
 */

package columbus.java.asg.struc;

import columbus.java.asg.*;
import columbus.java.asg.base.Named;

/**
 * Interface Package, which represents the {@link columbus.java.asg.struc.Package Package} node.
 * @columbus.node (missing)
 * @columbus.attr qualifiedName (String) : (missing)
 * @columbus.edge hasCompilationUnits ({@link columbus.java.asg.struc.CompilationUnit CompilationUnit}, multiple) : (missing)
 * @columbus.edge isInModule ({@link columbus.java.asg.struc.Module Module}, multiple) : (missing)
 */
public interface Package extends Named, Scope {

	/**
	 * Gives back the {@link columbus.java.asg.struc.Package#attributeQualifiedName qualifiedName} of the node.
	 * @return Returns with the qualifiedName.
	 */
	public String getQualifiedName();

	/**
	 * Gives back the Key of {@link columbus.java.asg.struc.Package#attributeQualifiedName qualifiedName} of the node.
	 * @return Returns with the Key of the qualifiedName.
	 */
	public int getQualifiedNameKey();

	/**
	 * Sets the {@link columbus.java.asg.struc.Package#attributeQualifiedName qualifiedName} of the node.
	 * @param value The new value of the qualifiedName.
	 */
	public void setQualifiedName(String value);

	/**
	 * Gives back iterator for the {@link columbus.java.asg.struc.Package#edgeHasCompilationUnits hasCompilationUnits} edges.
	 * @return Returns an iterator for the hasCompilationUnits edges.
	 */
	public EdgeIterator<CompilationUnit> getCompilationUnitsIterator();

	/**
	 * Tells whether the node has {@link columbus.java.asg.struc.Package#edgeHasCompilationUnits hasCompilationUnits} edges or not.
	 * @return Returns true if the node doesn't have any hasCompilationUnits edge.
	 */
	public boolean getCompilationUnitsIsEmpty();

	/**
	 * Gives back how many {@link columbus.java.asg.struc.Package#edgeHasCompilationUnits hasCompilationUnits} edges the node has.
	 * @return Returns with the number of hasCompilationUnits edges.
	 */
	public int getCompilationUnitsSize();

	/**
	 * Adds a new {@link columbus.java.asg.struc.Package#edgeHasCompilationUnits hasCompilationUnits} edge to the node.
	 * @param id The end point of the new hasCompilationUnits edge.
	 */
	public void addCompilationUnits(int id);

	/**
	 * Adds a new {@link columbus.java.asg.struc.Package#edgeHasCompilationUnits hasCompilationUnits} edge to the node.
	 * @param node The end point of the new hasCompilationUnits edge.
	 */
	public void addCompilationUnits(CompilationUnit node);

	/**
	 * Gives back iterator for the {@link columbus.java.asg.struc.Package#edgeIsInModule isInModule} edges.
	 * @return Returns an iterator for the isInModule edges.
	 */
	public EdgeIterator<Module> getIsInModuleIterator();

	/**
	 * Tells whether the node has {@link columbus.java.asg.struc.Package#edgeIsInModule isInModule} edges or not.
	 * @return Returns true if the node doesn't have any isInModule edge.
	 */
	public boolean getIsInModuleIsEmpty();

	/**
	 * Gives back how many {@link columbus.java.asg.struc.Package#edgeIsInModule isInModule} edges the node has.
	 * @return Returns with the number of isInModule edges.
	 */
	public int getIsInModuleSize();

	/**
	 * Adds a new {@link columbus.java.asg.struc.Package#edgeIsInModule isInModule} edge to the node.
	 * @param id The end point of the new isInModule edge.
	 */
	public void addIsInModule(int id);

	/**
	 * Adds a new {@link columbus.java.asg.struc.Package#edgeIsInModule isInModule} edge to the node.
	 * @param node The end point of the new isInModule edge.
	 */
	public void addIsInModule(Module node);

}

